### PR TITLE
Dk use iodata in iso 8601 and json serialization

### PR DESCRIPTION
**Pull request by dkuku**
continues on top of #14130 the changes in iso.ex are in the previous pr.
Main changes are the use of iodata in `iso8601` functions for date related modules that are used in  json serialization.
It also exposes the `Calendar.ISO.microseconds_to_iodata(ms, p)` as public because its the same logic as in the iso module that I refactored.

### CODE REVIEW

File: lib/elixir/lib/calendar/date.ex [hunk 1]
```diff
@@ -409,15 +409,24 @@ defmodule Date do
...
-  def to_iso8601(%{calendar: Calendar.ISO} = date, format) when format in [:basic, :extended] do
+  def to_iso8601(date, format) do
+    to_iso8601_iodata(date, format)
+    |> IO.iodata_to_binary()
+  end
+
+  @spec to_iso8601_iodata(Calendar.date(), :extended | :basic) :: iodata
+  def to_iso8601_iodata(date, format \\ :extended)
+
+  def to_iso8601_iodata(%{calendar: Calendar.ISO} = date, format)
```
Comment (josevalim) [inline lib/elixir/lib/calendar/date.ex:420]: Let’s not expose these functions from the date/time/naivedatetime/datetime APIs yet, only use it internally.

**Comment by josevalim** [comment]
@dkuku the other PR has been merged. To keep things simpler, let's only change the JSON implementation, not the actual calendar types. This will require a longer conversation about how to deal with Duration.to_iodata in json, so let's skip that in this PR as well. In a nutshell:

1. Use iodata for JSON in time, date, ndt, and datetime
2. Use microseconds_to_iodata in duration

**Comment by dkuku** [comment]
I'm a bit confused about the suggested change to the JSON implementation. When you mentioned "Only change the JSON implementation," did you mean that I should move the logic for generating the ISO 8601 representation as iodata to the JSON.Encoder module, and keep the existing to_iso8601 functions in the calendar types (e.g., Date, Time, NaiveDateTime, DateTime) unchanged? I had this in mind when I wake up but there are multiple function heads depending on the calendar used.

**Comment by josevalim** [comment]
> I'm a bit confused about the suggested change to the JSON implementation. When you mentioned "Only change the JSON implementation," did you mean that I should move the logic for generating the ISO 8601 representation as iodata to the JSON.Encoder module, and keep the existing to_iso8601 functions in the calendar types (e.g., Date, Time, NaiveDateTime, DateTime) unchanged? I had this in mind when I wake up but there are multiple function heads depending on the calendar used.

Yes, exactly. Add a new clause that, if the calendar type is Calendar.ISO, we use the special path, otherwise we fallback to today's implementation.

**Comment by dkuku** [comment]
I refactored this pr. Only 2 files are changed now.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**